Bottom: 34f0b5f09ddebfe05486a1aea68e2cbc6456eeb8
Top:    f866b722631777bcee7996ad7df35281af283841
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-05-30 15:48:27 +0800

vhost_net: introduce get_tx_bufs()

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 7f3418ca..723b4e8 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -485,6 +485,45 @@ static bool vhost_exceeds_weight(int pkts, int total_len)
 	       pkts >= VHOST_NET_PKT_WEIGHT;
 }
 
+static int get_tx_bufs(struct vhost_net *net,
+		       struct vhost_net_virtqueue *nvq,
+		       struct vhost_used_elem *used,
+		       struct msghdr *msg,
+		       unsigned int *out, unsigned int *in,
+		       size_t *len, bool *busyloop_intr)
+{
+	struct vhost_virtqueue *vq = &nvq->vq;
+	int ret;
+
+	ret = vhost_net_tx_get_vq_desc(net, vq, used, vq->iov,
+				       ARRAY_SIZE(vq->iov),
+				       out, in, busyloop_intr);
+	if (ret)
+		goto err;
+
+	if (*in) {
+		vq_err(vq, "Unexpected descriptor format for TX: "
+			"out %d, int %d\n", *out, *in);
+		ret = -EFAULT;
+		goto err;
+	}
+
+	/* Sanity check */
+	*len = init_iov_iter(vq, &msg->msg_iter, nvq->vhost_hlen, *out);
+	if (*len == 0) {
+		vq_err(vq, "Unexpected header len for TX: "
+			"%zd expected %zd\n",
+			*len, nvq->vhost_hlen);
+		ret = -EFAULT;
+		goto err;
+	}
+
+	return 0;
+err:
+	*len = 0;
+	return ret;
+}
+
 /* Expects to be always run from workqueue - which acts as
  * read-size critical section for our kind of RCU. */
 static void handle_tx(struct vhost_net *net)
@@ -502,7 +541,6 @@ static void handle_tx(struct vhost_net *net)
 	};
 	size_t len, total_len = 0;
 	int err;
-	size_t hdr_size;
 	struct socket *sock;
 	struct vhost_net_ubuf_ref *uninitialized_var(ubufs);
 	bool zcopy, zcopy_used;
@@ -519,7 +557,6 @@ static void handle_tx(struct vhost_net *net)
 	vhost_disable_notify(&net->dev, vq);
 	vhost_net_disable_vq(net, vq);
 
-	hdr_size = nvq->vhost_hlen;
 	zcopy = nvq->ubufs;
 
 	for (;;) {
@@ -529,10 +566,11 @@ static void handle_tx(struct vhost_net *net)
 		if (zcopy)
 			vhost_zerocopy_signal_used(net, vq);
 
+<<<<<<< current
 		busyloop_intr = false;
-		head = vhost_net_tx_get_vq_desc(net, vq, vq->iov,
-						ARRAY_SIZE(vq->iov),
-						&out, &in, &busyloop_intr);
+		head = get_tx_bufs(net, vq, vq->iov,
+				   ARRAY_SIZE(vq->iov),
+				   &out, &in, &busyloop_intr);
 		/* On error, stop handling until the next kick. */
 		if (unlikely(head < 0))
 			break;
@@ -541,11 +579,17 @@ static void handle_tx(struct vhost_net *net)
 			if (unlikely(busyloop_intr)) {
 				vhost_poll_queue(&vq->poll);
 			} else if (unlikely(vhost_enable_notify(&net->dev, vq))) {
+=======
+		err = get_tx_bufs(net, nvq, &used, &msg, &out, &in, &len);
+		if (err == -ENOSPC) {
+			if (unlikely(vhost_enable_notify(&net->dev, vq))) {
+>>>>>>> patched
 				vhost_disable_notify(&net->dev, vq);
 				continue;
 			}
 			break;
 		}
+<<<<<<< current
 		if (in) {
 			vq_err(vq, "Unexpected descriptor format for TX: "
 			       "out %d, int %d\n", out, in);
@@ -560,6 +604,11 @@ static void handle_tx(struct vhost_net *net)
 			len, hdr_size);
 			break;
 		}
+=======
+		/* On error, stop handling until the next kick. */
+		if (unlikely(err < 0))
+			break;
+>>>>>>> patched
 
 		zcopy_used = zcopy && len >= VHOST_GOODCOPY_LEN
 				   && !vhost_exceeds_maxpend(net)
